Katello/katello
# Fixes #36897 - Hosts now consistently follow rules when inheriting at...

…tributes from their hostgroups

#### What are the changes introduced in this pull request?
Despite the simplicity of the bug, there were some big changes introduced to how we handle host inheritance from host groups. On host creation, a host should still inherit the lifecycle environment and the content view from its hostgroup; these two properties (among a few others) are referred to as 'content facets'. The new changes introduced are as follows:

- On creation, hosts which do not have certain content facets assigned will inherit these properties from their host group (if it exists).
- Hosts can be manually requested to have certain content facets differing from the host group on creation. These requests will now be honored.
- Hosts that are given either a CV or an LE at creation can now inherit the missing content facets from the host group.
- Hosts can be updated to have different CVs and LEs from their host group. Unlike before, no updates (either to the host group or the host) will modify these properties unless the CV or LE itself is directly updated.

#### Considerations taken when implementing this change?
There are a lot of ramifications to this change. Jeremy and I talked about the expected behavior for a while but there is likely stuff that we missed or were wrong about. I'd be happy to chat about this PR any time.

One big consideration is that updating host groups does not update the host content LE or CV. Is this a feature users need/want or are they indifferent?

#### What are the testing steps for this pull request?
1. Create two different LEs, each with their own CV.
2. Create a host group assigned to LE 1.
3. Create a host assigned to the host group and register it. Ensure that the CV and LE match the host group. I had to use a VM host to get the registration to work properly, but allegedly there is an easier way to do this?
4. On the host details page for your host, click the menu next to 'Content view details'. Select 'Edit content view assignment' and assign the other LE and CV. Upon saving, you should see the content view details window update.
5. Click 'Edit' in the upper corner to edit the host. Modify the comments field in 'Additional information' and submit the changes. The CV and LE selected before should not have changed.

That takes care of the bug portion of this PR. Due to the changes with how inheritance is handled, it would be beneficial to check the functionality of a few other styles of host:
- Is behavior normal for hosts without a hostgroup?
- Is behavior normal for hostgroups that don't have a content view assigned?
- Check that hostgroup changes do not change existing hosts. Is this ok?

You can take a look at `host_managed_extensions_test.rb` for the theory behind my testing. Please ensure that these tests are sound.
